jaxmarl/environments/hanabi/hanabi.py

```diff
@@ -241,63 +241,57 @@ def get_legal_moves(self, state: State) -> chex.Array:
-                # get occurrences of each rank
-                rank_counts = jnp.sum(card_counts, axis=0)
-                # check which colors/ranks in hand
-                colors_present = jnp.where(color_counts > 0, 1, 0)
-                ranks_present = jnp.where(rank_counts > 0, 1, 0)
-
-                valid_hints = jnp.concatenate([colors_present, ranks_present])
-                carry = (aidx + 1, other_hands)
-
-                return carry, valid_hints
-
-            _, valid_hints = lax.scan(
-                _get_hints_for_hand, (0, other_hands), None, self.num_agents - 1
+            all_player_hands = state.player_hands  # (num_players, hand_size, num_colors, num_ranks)
+
+            # first get all cards one is holding since only these are legally playable
+            my_hand = all_player_hands.at[aidx].get()  # (hand_size, num_colors, num_ranks)
+            holding_cards_idx = jax.vmap(lambda c: jnp.any(c))(my_hand)
+
+            # discard is legal when tokens are not full and when card is in hand
+            can_get_token = jnp.sum(state.info_tokens) < self.max_info_tokens
+            legal_discard_idx = holding_cards_idx * can_get_token
+            legal_moves = legal_moves.at[self.discard_action_range].set(legal_discard_idx)
+
+            # play is legal for cards in hand - if empty card, not legal.
+            legal_moves = legal_moves.at[self.play_action_range].set(holding_cards_idx)
+
+            # hints depend on cards held by other players and not our own
+            other_players_hands = jnp.delete(
+                all_player_hands, aidx, axis=0, assume_unique_indices=True
             )
-            # make other player positions relative to current player
-            valid_hints = jnp.roll(valid_hints, -aidx, axis=0)
-            # include valid hints in legal moves
-            num_hints = (self.num_agents - 1) * (self.num_colors + self.num_ranks)
-            valid_hints = jnp.concatenate(valid_hints, axis=0)
-            info_tokens_available = jnp.sum(info_tokens) != 0
-            valid_hints *= info_tokens_available
-            legal_moves = legal_moves.at[move_idx : move_idx + num_hints].set(
-                valid_hints
+            # adjust to have relative positions
+            other_players_hands = jnp.roll(
+                other_players_hands, -aidx, axis=0
             )
 
-            # only enable noop if not current player
+            # cards can be hinted only if info tokens are available
+            info_tokens_available = jnp.sum(state.info_tokens) > 0
+
+            # get all the colors that can be hinted
+            def _hintable_colors(hand):
+                # Hand: (num_cards, num_colors, num_ranks)
+                card_colors = jnp.sum(hand, axis=2)
+                hintable_colors = card_colors.any(axis=0)
+                return hintable_colors
+
+            legal_color_hints = jax.vmap(_hintable_colors)(other_players_hands).ravel()
```

Comment:
This could also go without vmap probably, but this seemed like a natural way to implement it.